### PR TITLE
Feature/446 refactor job parameters

### PR DESCRIPTION
**Pull request by jozefbakus**
Implements: #446

### CODE REVIEW

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala [hunk 1]
```diff
@@ -37,48 +34,59 @@ object JobTemplateResolutionUtil {
...
+  private def mergeJobParameters(primary: JobDefinitionParameters, secondary: JobTemplateParameters): JobInstanceParameters = {
+    (primary, secondary) match {
+      case (definitionParams: SparkDefinitionParameters, templatePrams: SparkTemplateParameters) =>
+        mergeSparkParameters(definitionParams, templatePrams)
+      case (definitionParams: HyperdriveDefinitionParameters, templatePrams: SparkTemplateParameters) =>
+        mergeSparkAndHyperdriveParameters(definitionParams, templatePrams)
+      case (definitionParams: ShellDefinitionParameters, templatePrams: ShellTemplateParameters) =>
+        mergeShellParameters(definitionParams, templatePrams)
+      case _ =>
+        throw new IllegalArgumentException("Couldn't not mix different job types.")
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala]: Typo: `Couldn't not` -> `Could not`

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala [hunk 2]
```diff
@@ -37,48 +34,59 @@ object JobTemplateResolutionUtil {
...
+    }
   }
 
-  private def mergeMapsOfStrings(primary: Map[String, String], secondary: Map[String, String]) =
-    secondary ++ primary
-
-  private def mergeMapsOfLists(primary: Map[String, List[String]], secondary: Map[String, List[String]]) =
-    mergeMaps(primary, secondary, (_: String, first: List[String], second: List[String]) => second ++ first)
+  private def mergeSparkAndHyperdriveParameters(definitionParams: HyperdriveDefinitionParameters, templatePrams: SparkTemplateParameters): SparkInstanceParameters = {
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala]: Typo: `templatePrams` -> `templateParams`

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/DagInstanceService.scala [hunk 3]
```diff
@@ -54,9 +54,9 @@ class DagInstanceServiceImpl(override val jobTemplateService: JobTemplateService
     val now = LocalDateTime.now()
     val finished = if (skip) Some(now) else None
     resolvedJobDefinitions.map { resolvedJobDefinition =>
-      val jobParameters = resolvedJobDefinition.jobType match {
-        case JobTypes.Spark => SparkParameters(resolvedJobDefinition.jobParameters)
-        case JobTypes.Shell => ShellParameters(resolvedJobDefinition.jobParameters)
+      val jobParameters = resolvedJobDefinition.jobParameters.jobType match {
+        case JobTypes.Spark => resolvedJobDefinition.jobParameters
+        case JobTypes.Shell => resolvedJobDefinition.jobParameters
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/DagInstanceService.scala]: Isn't this now equivalent to `val jobParameters = resolvedJobDefinition.jobParameters`?

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowValidationService.scala [hunk 4]
```diff
@@ -141,6 +138,30 @@ class WorkflowValidationServiceImpl @Inject()(override val workflowRepository: W
     }
   }
 
+  private def validateJobParameters(originalJobParameters: JobDefinitionParameters, updatedJobParameters: JobDefinitionParameters): Seq[Boolean] = {
+    (originalJobParameters, updatedJobParameters) match {
+      case (original: SparkDefinitionParameters, updated: SparkDefinitionParameters) =>
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowValidationService.scala]: Can't you just write `original.equals(updated)`?

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala [hunk 5]
```diff
@@ -37,48 +34,59 @@ object JobTemplateResolutionUtil {
...
+      appArguments = mergeLists(definitionParams.appArguments, templatePrams.appArguments),
+      additionalJars = mergeLists(definitionParams.additionalJars, templatePrams.additionalJars),
+      additionalFiles = mergeLists(definitionParams.additionalFiles, templatePrams.additionalFiles),
+      additionalSparkConfig = mergeMapsOfStrings(definitionParams.additionalSparkConfig, templatePrams.additionalSparkConfig)
+    )
+  }
 
-  private def mergeSortedMapEntries(key: String, firstValue: String, secondValue: String) = {
-    if (KeysToMerge.contains(key)) {
-      s"$secondValue$MergedValuesSeparator$firstValue".trim
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala]: Looks like I've never written a test for this case. But in case the key is `spark.executor.extraJavaOptions` or `spark.driver.extraJavaOptions`, this additional logic applies. Please add it back (for additionalSparkConfig)

File: src/main/scala/za/co/absa/hyperdrive/trigger/Application.scala [hunk 6]
```diff
@@ -87,12 +90,58 @@ object ObjectMapperSingleton {
     }
   }
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/Application.scala]: It might be worth now to refactor the `ObjectMapperSingleton` into its own file

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/JobDefinitionParameters.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+  formConfig: FormConfig = FormConfigs.Spark,
+  jobJar: Option[String],
+  mainClass: Option[String],
+  appArguments: List[String] = List.empty[String],
+  additionalJars: List[String] = List.empty[String],
+  additionalFiles: List[String] = List.empty[String],
+  additionalSparkConfig: Map[String, String] = Map.empty[String, String]
+) extends JobDefinitionParameters
+
+case class HyperdriveDefinitionParameters(
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/JobDefinitionParameters.scala:37]: I would extract common fields with the `SparkDefinitionParameters` to a trait. For `HyperdriveDefinitionParameters`, jobJar and mainClass could be None

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/JobTemplateParameters.scala [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+import play.api.libs.json.{Format, JsResult, JsValue, Json, OFormat}
+import za.co.absa.hyperdrive.trigger.models.enums.JobTypes
+import za.co.absa.hyperdrive.trigger.models.enums.JobTypes.JobType
+
+sealed trait JobTemplateParameters {
+  val jobType: JobType
+}
+
+case class SparkTemplateParameters(
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/models/JobTemplateParameters.scala:27]: I think I'd add a trait to connect `SparkInstanceParameters`, `SparkDefinitionParameters` and `SparkTemplateParameters`, because they are all dependent on each other. E.g. if we add back the deployment mode, we'd have to change three places.

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala [hunk 9]
```diff
@@ -57,146 +55,246 @@ class JobTemplateResolutionUtilTest extends FlatSpec with Matchers {
...
+    val bothUndefined = JobTemplateResolutionUtil.resolveDagDefinitionJoined(
+      createDagDefinitionJoined(createJobDefinition().copy(jobTemplateId = jobTemplateUndefined.id, jobParameters = sparkJobParametersUndefined)),
+      Seq(jobTemplateUndefined)
+    )
+    val templateDefined = JobTemplateResolutionUtil.resolveDagDefinitionJoined(
+      createDagDefinitionJoined(createJobDefinition().copy(jobTemplateId = jobTemplateDefined.id, jobParameters = sparkJobParametersUndefined)),
+      Seq(jobTemplateDefined)
+    )
+    val jobDefined = JobTemplateResolutionUtil.resolveDagDefinitionJoined(
+      createDagDefinitionJoined(createJobDefinition().copy(jobTemplateId = jobTemplateUndefined.id, jobParameters = sparkJobParametersDefined)),
```
Comment (kevinwallimann) [inline src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala:160]: should it not be `jobTemplateDefined.id`?

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala [hunk 10]
```diff
@@ -57,146 +55,246 @@ class JobTemplateResolutionUtilTest extends FlatSpec with Matchers {
...
+      additionalFiles = List.empty[String],
+      additionalSparkConfig = Map.empty[String, String]
+    )
+    val sparkTemplateParametersDefined = SparkTemplateParameters(
+      jobJar = Option("templateJar.jar"),
+      mainClass = Option("templateClass"),
+      appArguments = List("templateAppArgument1", "templateAppArgument2", "appArgument"),
+      additionalJars = List("templateJar1", "templateJar2", "jar"),
+      additionalFiles = List("templateFile1", "templateFile2", "file"),
+      additionalSparkConfig = Map("templateKey1" -> "templateValue1", "templateKey2" -> "templateValue2", "key" -> "value2")
```
Comment (kevinwallimann) [inline src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala]: I would use some shared keys as well, e.g. "sharedKey1" -> "templateValueSharedKey1", and the same for the jobParameters above, e.g. "sharedKey1" -> "jobValueSharedKey1". And also, as mentioned in another comment, use "spark.executor.extraJavaOptions" and "spark.driver.extraJavaOptions" as one of the shared keys (but maybe in a separate test)

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala [hunk 11]
```diff
@@ -57,146 +55,246 @@ class JobTemplateResolutionUtilTest extends FlatSpec with Matchers {
...
+    bothScriptsDefined.head.jobParameters.asInstanceOf[SparkInstanceParameters].mainClass shouldBe sparkTemplateParametersDefined.mainClass.get
+    bothScriptsDefined.head.jobParameters.asInstanceOf[SparkInstanceParameters].appArguments should contain theSameElementsAs hyperdriveJobParametersDefined.appArguments ++ sparkTemplateParametersDefined.appArguments
+    bothScriptsDefined.head.jobParameters.asInstanceOf[SparkInstanceParameters].additionalJars should contain theSameElementsAs hyperdriveJobParametersDefined.additionalJars ++ sparkTemplateParametersDefined.additionalJars
+    bothScriptsDefined.head.jobParameters.asInstanceOf[SparkInstanceParameters].additionalFiles should contain theSameElementsAs hyperdriveJobParametersDefined.additionalFiles ++ sparkTemplateParametersDefined.additionalFiles
+    bothScriptsDefined.head.jobParameters.asInstanceOf[SparkInstanceParameters].additionalSparkConfig should contain theSameElementsAs Map("templateKey1" -> "templateValue1", "templateKey2" -> "templateValue2", "jobKey1" -> "jobValue1", "jobKey2" -> "jobValue2", "key" -> "value1")
   }
 
-  it should "in key-value pairs, concatenate the values if the key is extraJavaOptions" in {
```
Comment (kevinwallimann) [inline src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala:173]: Ah, here it was. Please adapt this test case

File: ui/src/app/components/admin/job-templates/job-templates-home/job-templates-home.component.html [hunk 12]
```diff
@@ -48,13 +48,8 @@
       </app-string-filter>
     </clr-dg-filter>
   </clr-dg-column>
-  <clr-dg-column [clrDgField]="jobTemplateColumns.JOB_TYPE">
+  <clr-dg-column>
     <ng-container *clrDgHideableColumn="{hidden: false}">Job type</ng-container>
-    <clr-dg-filter [clrDgFilter]="jobTypeFilter">
```
Comment (kevinwallimann) [inline ui/src/app/components/admin/job-templates/job-templates-home/job-templates-home.component.html:53]: Why did you remove this? Is it because it's a nested property now?

File: ui/src/app/stores/workflows/workflows.effects.ts [hunk 13]
```diff
@@ -247,50 +248,64 @@ export class WorkflowsEffects {
...
-            result.name,
-            result.isActive,
-            result.project,
-            result.created,
-            result.updated,
-            result.id,
-          );
-          this.toastrService.success(texts.CREATE_WORKFLOW_SUCCESS_NOTIFICATION);
-          this.router.navigateByUrl(absoluteRoutes.SHOW_WORKFLOW + '/' + workflow.id);
+      return this.workflowService.getJobTemplates().pipe(
```
Comment (kevinwallimann) [inline ui/src/app/stores/workflows/workflows.effects.ts:251]: This is quite a large effort just to get 'Spark' or 'Shell' :) But ok, since creation/editing of workflows is rare, I'm ok with this.

File: src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql [hunk 14]
```diff
@@ -0,0 +1,103 @@
+/*
```
Comment (kevinwallimann) [inline src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:1]: What about the db_script_latest.sql?

File: src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+ * Copyright 2018 ABSA Group Limited
```
Comment (kevinwallimann) [inline src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:2]: You could already add the next script for v0.4.5 to drop the variables, maps, key_value_pairs columns, but not add it in the changelog, but that's up to you. Either way someone has to do it in the next version

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala (hunk 9 above)
Comment (jozefbakus) [reply src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala:160]: No it this case only job params are defined.
bothUndefined - no job/template parameters
templateDefined - only template params defined
jobDefined - only job params defined
bothScriptsDefined - both defined

File: ui/src/app/components/admin/job-templates/job-templates-home/job-templates-home.component.html (hunk 12 above)
Comment (jozefbakus) [reply ui/src/app/components/admin/job-templates/job-templates-home/job-templates-home.component.html:53]: Yes because of the json

File: ui/src/app/stores/workflows/workflows.effects.ts (hunk 13 above)
Comment (jozefbakus) [reply ui/src/app/stores/workflows/workflows.effects.ts:251]: Yes, but I didn't have other option :/

File: src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql (hunk 15 above)
Comment (jozefbakus) [reply src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:2]: I will do it in its own PR

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala (hunk 1 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala]: Done

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala (hunk 2 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala]: Done

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/DagInstanceService.scala (hunk 3 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/DagInstanceService.scala]: Yes, fixed

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowValidationService.scala (hunk 4 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/WorkflowValidationService.scala]: Yes, fixed.

File: src/main/scala/za/co/absa/hyperdrive/trigger/Application.scala (hunk 6 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/Application.scala]: Yes, fixed.

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/JobDefinitionParameters.scala (hunk 7 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/JobDefinitionParameters.scala:37]: I like it now this way. I will update it in the next story.

File: src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql (hunk 14 above)
Comment (jozefbakus) [reply src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:1]: Oh, fixed

File: ui/src/app/components/admin/job-templates/job-templates-home/job-templates-home.component.html (hunk 12 above)
Comment (kevinwallimann) [reply ui/src/app/components/admin/job-templates/job-templates-home/job-templates-home.component.html:53]: ok

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala (hunk 9 above)
Comment (kevinwallimann) [reply src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala:160]: ah yes sorry

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala (hunk 5 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtil.scala]: Done

File: src/main/scala/za/co/absa/hyperdrive/trigger/models/JobTemplateParameters.scala (hunk 8 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/models/JobTemplateParameters.scala:27]: Let's me do it in next iteration.

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala (hunk 10 above)
Comment (jozefbakus) [reply src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala]: Done

File: src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala (hunk 11 above)
Comment (jozefbakus) [reply src/test/scala/za/co/absa/hyperdrive/trigger/api/rest/utils/JobTemplateResolutionUtilTest.scala:173]: Done

File: src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+alter table job_template rename column key_value_pairs to key_value_pairs_old;
+
+update job_template
+set job_parameters = job_parameters::jsonb || ('{"jobType":"' || job_type || '"}' )::jsonb;
+
+alter table job_template rename column job_type to job_type_old;
+alter table job_template alter column job_type_old drop not null;
+
+update workflow_history
+set workflow = jsonb_set(workflow::jsonb, array['dagDefinitionJoined', 'jobDefinitions'], updated.updatedJobDefinitions)
```
Comment (kevinwallimann) [inline src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:64]: I might have overlooked something, but according to the `db_script.sql`, `workflow_history.workflow` is still of type `varchar`. Does this work to set it with `jsonb_set`?

Comment (jozefbakus) [reply src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:64]: Yes it does, because workflow is a valid json

File: src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        jsonb_array_elements(workflow::jsonb -> 'dagDefinitionJoined' -> 'jobDefinitions') -> 'jobType' -> 'name' as jobTypeName,
+        jsonb_array_elements(workflow::jsonb -> 'dagDefinitionJoined' -> 'jobDefinitions') -> 'jobTemplateId' as templateId
+      from workflow_history
+    ) as toUpdate
+    left outer join job_template as jobTemplate
+    on toUpdate.templateId::bigint = jobTemplate.id
+  ) as toUpdateWithFormConfig
+  group by toUpdateWithFormConfig.historyId
+) as updated
+where updated.historyId=workflow_history.id;
```
Comment (kevinwallimann) [inline src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:109]: Btw, you could make this script idempotent by adding this line
```
  and workflow_history.workflow::jsonb -> 'dagDefinitionJoined' -> 'jobDefinitions' -> 'jobParameters' ? 'maps';
```
It shouldn't be a problem in production, but during testing, I (accidentally) reran the update query twice and then the jobDefinitions of all workflow histories were set to `[null]`

File: ui/src/app/models/formConfing.model.ts [hunk 18]
```diff
@@ -13,12 +13,12 @@
  * limitations under the License.
  */
 
-package za.co.absa.hyperdrive.trigger.models
+export type FormConfig = {
```
Comment (kevinwallimann) [inline ui/src/app/models/formConfing.model.ts]: Filename typo: `formConfing.model.ts` -> `formConfig.model.ts`

File: src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql (hunk 17 above)
Comment (jozefbakus) [reply src/main/resources/db_scripts/liquibase/v0.4.4.refactor-job-parameters.sql:109]: Done, but as discussed it has to go deeper into the script because 'jobDefinitions' is an array

File: ui/src/app/models/formConfing.model.ts (hunk 18 above)
Comment (jozefbakus) [reply ui/src/app/models/formConfing.model.ts]: Done

**Review by kevinwallimann — APPROVED** [review]
